egs/librispeech/ASR/local/prepare_lang.py

```diff
@@ -0,0 +1,417 @@
+    the beginning and end of the word.
+
+    Args:
+      lexicon:
+        The input lexicon. See also :func:`read_lexicon`
+      phone2id:
+        A dict mapping phones to IDs.
+      word2id:
+        A dict mapping words to IDs.
+      sil_phone:
+        The silence phone.
+      sil_prob:
+        The probability for adding a silence at the beginning and end
+        of the word.
+      need_self_loops:
+        If True, add self-loop to states with non-epsilon output symbols
+        on at least one arc out of the state.
+    Returns:
+      Return an instance of `k2.Fsa` representing the given lexicon.
+    """
+    assert sil_prob > 0.0 and sil_prob < 1.0
+    # CAUTION: we use score, i.e, negative cost.
+    sil_score = math.log(sil_prob)
+    no_sil_score = math.log(1.0 - sil_prob)
+
+    start_state = 0
+    loop_state = 1  # words enter and leave from here
+    sil_state = 2  # words terminate here when followed by silence; this state
+    # has a silence transition to loop_state.
+    next_state = 3  # the next un-allocated state, will be incremented as we go.
+    arcs = []
+
+    assert phone2id["<eps>"] == 0
+    assert word2id["<eps>"] == 0
+
+    eps = 0
+
+    sil_phone = phone2id[sil_phone]
+
+    arcs.append([start_state, loop_state, eps, eps, no_sil_score])
+    arcs.append([start_state, sil_state, eps, eps, sil_score])
+    arcs.append([sil_state, loop_state, sil_phone, eps, 0])
+
+    for word, prons in lexicon:
+        assert len(prons) > 0, f"{word} has no pronunciations"
+        cur_state = loop_state
+
+        word = word2id[word]
+        prons = [phone2id[i] for i in prons]
+
+        for i in range(len(prons) - 1):
+            if i == 0:
+                arcs.append([cur_state, next_state, prons[i], word, 0])
+            else:
+                arcs.append([cur_state, next_state, prons[i], eps, 0])
+
+            cur_state = next_state
+            next_state += 1
+
+        # now for the last phone of this word
+        # It has two out-going arcs, one to the loop state,
+        # the other one to the sil_state.
+        i = len(prons) - 1
+        w = word if i == 0 else eps
+        arcs.append([cur_state, loop_state, prons[i], w, no_sil_score])
+        arcs.append([cur_state, sil_state, prons[i], w, sil_score])
+
+    if need_self_loops:
+        disambig_phone = phone2id["#0"]
+        disambig_word = word2id["#0"]
+        arcs = add_self_loops(
+            arcs,
+            disambig_phone=disambig_phone,
+            disambig_word=disambig_word,
+        )
+
+    final_state = next_state
+    arcs.append([loop_state, final_state, -1, -1, 0])
+    arcs.append([final_state])
+
+    arcs = sorted(arcs, key=lambda arc: arc[0])
+    arcs = [[str(i) for i in arc] for arc in arcs]
+    arcs = [" ".join(arc) for arc in arcs]
+    arcs = "\n".join(arcs)
+
+    fsa = k2.Fsa.from_str(arcs, acceptor=False)
+    return fsa
+
+
+def main():
+    out_dir = Path("data/lang")
```

Comment:
If these constants became default arguments with `argparse` or `click`, then this is a pretty cool example of a Python “library-module+script” (you can both import from it and run it as an executable). Do you think that it needs to stay in local (i.e. do you expect other recipes will deviate from this workflow?)

Reply:
I agree.

Reply:
Guys, let's be cautious about putting things in the shared global space until we feel the recipes have stabilized.
I don't want to build up too much shared infrastructure while there are questions about whether something is a stable pattern yet.  It's always easy to put something in the icefall library, but taking things out can be hard.

Reply:
.. also, @pzelasko, I actually agree with the way @csukuangfj has done it here.

There may be some overlap, yes, but I think it's nice to have a separate script for turning the lexicon.txt into the lang dir.  This makes it easy to modify this script separately from the reading code (and we can know, for example, that the code responsible for reading the lexicon won't break when we change this).  I always prefer having a self-contained script, versus one that depends on a large library, because the self-contained script can be reasoned about in isolation, without worrying about which library version it depends on and what the interface of that library is.  So it is more self-documenting.

Also, at some point we should formalize the requirements on the input lexicon.txt file; and we'll eventually need a validation process for the lexicon.txt.  I'm thinking that we should have the concept of a "dict dir" like in Kaldi, and later on we may choose to have the phones.txt be an optional extra input in that dict dir in case we need to sync phone-sets across lexicons that have some phones missing.   By keeping the "dict dir" and "lang dir" separate, we can clarify inputs vs. output, and make the task of validation more straightforward to understand.

Reply:
Alright, I see your point.